ERP integration guide (#260)
Trackunit/developer-hub: guides/integrations-connectors/custom-erp-integration-guide.md

- Fleet snapshot pages: Maximum 50 requests per second
- Single-element snapshot endpoint: Maximum 500 requests per second

To avoid hitting rate limits, it is important to schedule data retrieval efficiently and implement retry logic when encountering rate limit errors.

## Push rental ERP data from an ERP system into Trackunit

In modern fleet management, integrating rental ERP systems with telematics solutions like Trackunit is crucial for enhancing operational efficiency and data consistency. Trackunit’s Rental API enables rental companies to seamlessly transfer key data—such as customers and contracts—from their Rental ERP system into Trackunit Manager, where users can monitor and manage their assets in real-time.

> 📘 IrisX Subscription needed
> 
> The Rental ERP API is only available for IrisX enabled customers. Learn more about the [IrisX subscription](https://developers.trackunit.com/docs/irisx-overview).

### Step 1: Understand and map concepts and data

> ![Trackunit Rental Data Model](https://cdn.statically.io/gh/trackunit/developer-hub/master/guides/integrations-connectors/erp-rental-data-model.png)

Reviewer: Is it intentional that the link is hardcoded to the statically cdn? Wouldn't a relative path make more sense if we change CDN in the future?

> ![Trackunit Rental Data Model](https://cdn.statically.io/gh/trackunit/developer-hub/master/guides/integrations-connectors/erp-rental-data-model.png)

Reviewer: This model contains most of the structure (concepts) but it is not a complete picture of all attributes. Should that be reflected in the text?

The Trackunit data model consists of the following concepts for rental ERP data:

#### Data points connected to an asset:

- **Identifier Type:** It is possible to match with asset in Trackunit Iris based on their asset id, serial number or external reference. This field combined with the identifier field is used when looking up an asset.

> - **Identifier:** The identifier used for matching the asset that the contract information should be assigned to.

Reviewer: ```suggestion
- **Identifier:** The identifier used for matching the asset that the contract information and rental status should be assigned to.
```

> - **Rental status:** The rental status can be used to locate possible available assets accross all depots. Available types are : 

Reviewer: The available types are available in the endpoint documentation. Do we want to maintain this list here as well?

    - ON_RENT, 
    - PICK_UP_READY, 
    - AVAILABLE, 
    - RETURNED, 
    - IN_REPAIR, 
    - TRANSFER, 
    - OTHER, 
    - OFF_RENT, 
    - NOT_ON_CONTRACT, 
    - RESERVED. 
    - New types may be added.
- **Owning depot:** Depots created in Trackunit Iris can be linked to assets to have an overview of all assets on a depot level.
- **ExternalReference** represents an identifier from the ERP system and has to be unique.

![Rental ERP data on assets](https://cdn.statically.io/gh/trackunit/developer-hub/master/guides/integrations-connectors/erp-data-asset.png)

#### Contracts
The contracts can be used for assigning an unassigning assets to and from customers which also makes them available in the customer portal. Contracts can have relations to many contract items, which hold details about specific rental periods.

- **Associated customer**
- **External Reference** to the contract in an external system, i.e., an identifier of the item. Must be unique.

> - **Contract status:**

Reviewer: The available types are available in the endpoint documentation. Do we want to maintain this list here as well?

    - OPEN
    - RESERVED
    - CANCELLED
    - COMPLETED
    - New values may be added.
- **Start date** and **end date** of the contract

![Rental ERP data on contracts](https://cdn.statically.io/gh/trackunit/developer-hub/master/guides/integrations-connectors/erp-data-contract.png)

#### Data points connected to a contract item:

- **Rental period:** Determines when an asset should be available for a customer.
    - OnRent Date
    - OffRate Date
- **Rental Rate:** Rental rates can contain information about price for renting assets.
- **Hiring depot:** When sharing data about hiring depot as part of the rental contract your customers can get access to data about which depot has rented an asset to them.
- **Delivery / pickup address:** By sharing information about delivery/pickup address it will be possible to see where and when assets are expected to be delivered and/or picked up.

> 

Reviewer: ```suggestion
- **Delivery / pickup date:** share information about when the delivery and pickup of the asset can be expected.

```

#### Customers
By providing data about customers you will be able to share data about the asset with customers through customer portals. The information about customers and the asset they have access to can also be used to filter assets and get an overview of what assets a customer has access to.

- **externalReference:** External identifier of the customer in the Iris platform
- **Type of customer**
    - Contractor
    - Dealer
    - Distributor
    - OEM
    - Rental
    - Service
    - Other

> 💡 Note
> 

> > Once you create Customers through the ERP integration, the Customer feature in Manager will no longer allow for manual creation to preserve system integrity. 

Reviewer: ```suggestion
> Once the Rental API is enabled the Customer feature in Manager will no longer allow for manual creation or editing to preserve system integrity. 
```

### Step 2: Extending the data model for custom needs

Custom fields provide a way to define new fields in the Trackunit data model. Allowing to extend and customize Trackunit Manager. Currently we support extending the data model of assets, accounts, groups, sites, customers and rental contracts with new fields.

- To define a custom field, you first need to [add a definition using the Custom Fields Definitions API](https://developers.trackunit.com/reference/custom-field-definitions).
- Use the [Custom Fields Values API](https://developers.trackunit.com/reference/custom-field-values)to set values on your defined custom fields:
    - Specify the `entityId` for your query. Depending on which domain type you are interested in, this can be an assetId, accountId, groupId, siteId, customerId or rentalContractId.

> 💡 Custom Field Examples
> 
> Specifically for rental contracts, you can think of custom fields to store any additional data that is not accounted for in Trackunit’s rental data model - for example: “customer project code” or “service reference”.

### Step 3: Synchronizing ERP data via the Rental API

#### Preparation of data

> - Get an overview of all available assets in Trackunit and their unique identifiers like assetID and serial numbers (PIN / VIN etc) via the [Asset API - get assets endpoint](https://developers.trackunit.com/reference/getassets_v1)

Reviewer: ```suggestion
- Get an overview of all available assets in Trackunit and their unique identifiers like asset ID and serial numbers (PIN / VIN etc) via the [Asset API - get assets endpoint](https://developers.trackunit.com/reference/getassets_v1)
```

- Prepare assets by setting **ExternalReference** via the [Asset API - update asset endpoint](https://developers.trackunit.com/reference/patchasset_v1)
- If you wish to use the “owning depot” and “hiring depot” features, then use the [Sites API to get all sites](https://developers.trackunit.com/reference/getsites) created in Trackunit and add new sites by using the [create site endpoint](https://developers.trackunit.com/reference/createsite). Remember to choose the type “depot”.

#### Synchronize ERP data

> Use the [Rental API - Synchronize rental data endpoint](https://developers.trackunit.com/reference/synchronize) to start pushing rental ERP data to Trackunit.

Reviewer: ```suggestion
Use the [Rental API - Synchronize rental data endpoint](https://developers.trackunit.com/reference/synchronize) to start pushing rental ERP data to Trackunit.
Data can be pushed for one asset per request.
```
